libyul/optimiser/FunctionCallFinder.cpp

```diff
@@ -16,23 +16,46 @@
 */
 
 #include <libyul/optimiser/FunctionCallFinder.h>
+
+#include <libyul/optimiser/ASTWalker.h>
 #include <libyul/AST.h>
 
 using namespace solidity;
 using namespace solidity::yul;
 
-std::vector<FunctionCall*> FunctionCallFinder::run(Block& _block, YulString _functionName)
+namespace
+{
+template<typename Base, typename ResultType>
+class MaybeConstFunctionCallFinder: Base
 {
-	FunctionCallFinder functionCallFinder(_functionName);
-	functionCallFinder(_block);
-	return functionCallFinder.m_calls;
+public:
+	using MaybeConstBlock = std::conditional_t<std::is_const_v<ResultType>, Block const, Block>;
+	static std::vector<ResultType*> run(MaybeConstBlock& _block, YulString const _functionName)
+	{
+		MaybeConstFunctionCallFinder functionCallFinder(_functionName);
+		functionCallFinder(_block);
+		return functionCallFinder.m_calls;
+	}
+private:
+	explicit MaybeConstFunctionCallFinder(YulString const _functionName): m_functionName(_functionName), m_calls() {}
+	using Base::operator();
+	void operator()(ResultType& _functionCall) override
```

Comment:
I was just thinking about something like this - but maybe its really not worth it:

```c++
template<typename Base>
struct FunctionCallFinder: Base
{
  typename FunctionCall ResultType;
  void operator()(FunctionCall& _functionCall)
  {
    Base::operator()(_functionCall);
    if (_functionCall.functionName.name == m_functionName)
      m_calls.emplace_back(&_functionCall);
  }
private:
	YulString m_functionName;
	std::vector<ResultType*> m_calls;
};

template<typename Base>
struct SomethingElseFinder: Base
{
  typename SomethingElse ResultType;
  void operator()(FunctionCall& _functionCall)
  {
    Base::operator()(_functionCall);
    ....
  }
private:
   ...
};

template<typename Base, template <typename T> class Policy>
class MaybeConstFinder: Policy<Base>
{
  static std::vector<Policy<Base>::ResultType*> run(..)
  {
  }
}

MaybeConstFinder<ASTModifier, FunctionCallFinder>::run(..)
MaybeConstFinder<ASTModifier, SomethingElseFinder >::run(..)
```

Reply:
I'm not sure this is worth generalizing. Probably better to leave it until we actually run into another case that would fit this pattern. I think that the "finders" we actually need in practice are more varied, like needing to visit multiple node types or run additional logic.

Reply:
yeah - I was also not sure whether this would make any sense.. but I also did not really check what things we are currently collecting where and how.. however, I think if we use similar concepts already it might be nice to unify them